src/software/thunderscope/robot_diagnostics/chicker.py

```diff
@@ -73,6 +109,82 @@ def __init__(self, proto_unix_io):
         self.geneva_value = 3
         self.power_value = 1
 
+    def send_command_and_timeout(self, command):
+        """
+        If buttons are enabled, sends a Kick command and disables buttons
+
+        Attaches a callback to re-enable buttons after 3 seconds
+
+        :param command: Command to send. One of ChickerCommandMode.KICK or ChickerCommandMode.CHIP
+        """
+        # if button is enabled
+        if self.kick_chip_buttons_enable:
+            # send kick primitive
+            self.send_command(command)
+            self.disable_kick_chip_buttons()
+
+            # set and start timer to re-enable buttons after 3 seconds
+            QTimer.singleShot(
+                3 * MILLISECONDS_PER_SECOND, self.enable_kick_chip_buttons
+            )
+
+    def disable_kick_chip_buttons(self):
+        """
+        Disables the buttons
+        """
+        self.kick_chip_buttons_enable = False
+
+    def enable_kick_chip_buttons(self):
+        """
+        If buttons should be enabled, enables them
+        """
+        if self.no_auto_selected:
+            self.kick_chip_buttons_enable = True
+
+    def set_should_enable_buttons(self, enable):
+        """
+        Changes if buttons are clickable or not based on boolean parameter
+
+        :param enable: boolean to indicate whether buttons should be made clickable or not
+        """
+        self.no_auto_selected = enable
+
+        if enable:
+            self.enable_kick_chip_buttons()
+        else:
+            self.disable_kick_chip_buttons()
+
+    def send_command(self, command):
+        """Sends a [auto]kick or [auto]chip primitive
+
+        :param command: enum int value to indicate what primitive to send
+        :returns: None
+
+        """
+
+        # gets slider values
+        geneva_value = self.geneva_slider.value()
+
+        power_value = self.power_slider.value()
+
+        power_control = PowerControl()
+        power_control.geneva_slot = geneva_value
+
+        # sends kick, chip, autokick, or autchip primitive
+        if command == ChickerCommandMode.KICK:
+            power_control.chicker.kick_speed_m_per_s = power_value
+        elif command == ChickerCommandMode.CHIP:
+            power_control.chicker.chip_distance_meters = power_value
+        elif command == ChickerCommandMode.AUTOKICK:
+            power_control.chicker.auto_chip_or_kick.autokick_speed_m_per_s = power_value
+        elif command == ChickerCommandMode.AUTOCHIP:
+            power_control.chicker.auto_chip_or_kick.autochip_distance_meters = (
+                power_value
+            )
+
+        # sends proto
+        self.proto_unix_io.send_proto(PowerControl, power_control)
```

Comment:
there is a dangerous bug here which I didn't realize. When we have KICK/CHIP (not autokick/autochip), we appear to be spamming KICK/CHIP commands to the robot (this isn't correct right @LiCody). The bug is related to the fact that robot_communication.py's sends primitives 60Hz and uses a non-blocking PrimitiveSet ThreadSafeBuffer which keeps sending out the cached-value if a new Primitive isn't available in the buffer

anyways my easy-fix/suggestion is that for KICK or CHIP commands is to set power_control.chicker.chip_distance_meters to 0 after sending the PowerControl proto

Reply:
done

Reply:
I'm not sure if I fully understand what the bug is. When I run the code it doesn't seem to spam KICK/CHIP (no auto) primitives... Are you on the latest commit on the branch?
Also, `power_control` is a local variable created at the start of every `send_command`, so setting its value to 0 shouldn't have any effect on future protos sent.